bin/generate_missing_mappings

```diff
@@ -0,0 +1,328 @@
+
+  def load_destination_taxonomy_ids(output_taxonomy)
+    YAML.load_file("data/integrations/#{output_taxonomy}/full_names.yml").each_with_object({}) do |category, hash|
+      hash[category["full_name"]] = category["id"]
+    end
+  end
+
+  def generate_mapping(source_category_id:, source_category_name:, index_name:,
+    destination_taxonomy_ids_by_full_name:
+  )
+    category_embedding = get_embeddings(source_category_name)
+    top_candidate = search_top_candidate(query_embedding: category_embedding, index_name:)
+    destination_category_id = destination_taxonomy_ids_by_full_name[top_candidate]
+
+    new_entry = {
+      "input" => { "product_category_id" => source_category_id },
+      "output" => { "product_category_id" => [destination_category_id.to_s] },
+    }
+
+    mapping_to_be_graded = {
+      from_category_id: source_category_id,
+      from_category: source_category_name,
+      to_category_id: destination_category_id.to_s,
+      to_category: top_candidate,
+    }
+
+    grading_result = grade_taxonomy_mapping(mapping_to_be_graded)
+
+    { new_entry: new_entry, mapping_to_be_graded: mapping_to_be_graded, grading_result: grading_result }
+  end
+
+  def get_embeddings(text)
+    with_retries do
+      response = @openai_client.embeddings(
+        parameters: {
+          model: EMBEDDING_MODEL,
+          input: text,
+        },
+      )
+      response.dig("data", 0, "embedding")
+    end
+  end
+
+  def search_top_candidate(query_embedding:, index_name:)
+    result = @qdrant_client.points.search(
+      collection_name: index_name,
+      vector: query_embedding,
+      with_payload: true,
+      limit: 1,
+    )
+    result["result"].first["payload"][index_name]
+  end
+
+  def grade_taxonomy_mapping(mapping)
+    with_retries do
+      response = @openai_client.chat(
+        parameters: {
+          model: MAPPING_GRADER_GPT_MODEL,
+          messages: [
+            { role: "system", content: system_prompts_of_taxonomy_mapping_grader },
+            { role: "user", content: [mapping].to_json },
+          ],
+          temperature: 0,
+        },
+      )
+      JSON.parse(response.dig("choices", 0, "message", "content")).first["agree_with_mapping"]
+    end
+  end
+
+  def with_retries
+    retries = 0
+    begin
+      yield
+    rescue StandardError => e
+      retries += 1
+      if retries <= MAX_RETRIES
+        puts "Received error: #{e.message}. Retrying (#{retries}/#{MAX_RETRIES})..."
+        sleep(1)
+        retry
+      else
+        puts "Failed after #{MAX_RETRIES} retries."
+        raise
+      end
+    end
+  end
+
+  def write_disagree_messages(disagree_messages)
+    File.open("tmp/mapping_update_message.txt", "a") do |file|
+      file.puts "❗AI Grader disagrees with the following mappings:"
+      disagree_messages.each do |mapping|
+        mapping.each { |key, value| file.puts "#{key}:#{value}" }
+        file.puts
+      end
+    end
+  end
+
+  def system_prompts_of_taxonomy_mapping_grader
```

Comment:
Should this prompt have a promptfoo setup somewhere?

Reply:
Yea, we can establish a benchmark using promptfoo in the tools repository or another suitable location to evaluate and improve the prompt.